### ISSUE TITLE
bunzip regular seviri hrit segments

### ISSUE THREAD

**Issue by gerritholl** [pytroll/satpy#1975]
**Describe the bug**

https://github.com/pytroll/satpy/pull/1798 added bunzipping for seviri hirt prologue and epilogue files, but not for the actual segments

**To Reproduce**

My directory `/media/nas/x21308/scratch/SEVIRI/202201081200/...` has:

```
H-000-MSG4__-MSG4________-_________-EPI______-202201081200-__.bz2
H-000-MSG4__-MSG4________-HRV______-000001___-202201081200-__.bz2
H-000-MSG4__-MSG4________-HRV______-000002___-202201081200-__.bz2
H-000-MSG4__-MSG4________-HRV______-000003___-202201081200-__.bz2
H-000-MSG4__-MSG4________-HRV______-000004___-202201081200-__.bz2
H-000-MSG4__-MSG4________-HRV______-000005___-202201081200-__.bz2
H-000-MSG4__-MSG4________-HRV______-000006___-202201081200-__.bz2
H-000-MSG4__-MSG4________-HRV______-000007___-202201081200-__.bz2
H-000-MSG4__-MSG4________-HRV______-000008___-202201081200-__.bz2
H-000-MSG4__-MSG4________-HRV______-000009___-202201081200-__.bz2
...
H-000-MSG4__-MSG4________-WV_062___-000008___-202201081200-__.bz2
H-000-MSG4__-MSG4________-WV_073___-000001___-202201081200-__.bz2
H-000-MSG4__-MSG4________-WV_073___-000002___-202201081200-__.bz2
H-000-MSG4__-MSG4________-WV_073___-000003___-202201081200-__.bz2
H-000-MSG4__-MSG4________-WV_073___-000004___-202201081200-__.bz2
H-000-MSG4__-MSG4________-WV_073___-000005___-202201081200-__.bz2
H-000-MSG4__-MSG4________-WV_073___-000006___-202201081200-__.bz2
H-000-MSG4__-MSG4________-WV_073___-000007___-202201081200-__.bz2
H-000-MSG4__-MSG4________-WV_073___-000008___-202201081200-__.bz2
```

```python
from glob import glob
from satpy import Scene
from satpy.utils import debug_on; debug_on()
filenames = glob("/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4*")
sc = Scene(filenames=filenames, reader=["seviri_l1b_hrit"])
sc.load(["VIS006"])
```

**Expected behavior**

I expect loading to be successful.

**Actual results**

It's assigning the prologue and the epilogue but not the rest.

```
[DEBUG: 2022-01-12 15:11:55 : satpy.readers.yaml_reader] Reading ('/data/gholl/mambaforge/envs/py39/lib/python3.9/site-packages/satpy/etc/readers/seviri_l1b_hrit.yaml',)
/data/gholl/mambaforge/envs/py39/lib/python3.9/site-packages/satpy/readers/seviri_base.py:453: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  ('GsicsCalMode', np.bool),
/data/gholl/mambaforge/envs/py39/lib/python3.9/site-packages/satpy/readers/seviri_base.py:454: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  ('GsicsCalValidity', np.bool),
[DEBUG: 2022-01-12 15:11:55 : satpy.readers.yaml_reader] Assigning to seviri_l1b_hrit: ['/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-_________-PRO______-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-_________-EPI______-202201081200-__.bz2']
[INFO: 2022-01-12 15:11:55 : satpy.readers.utils] Using temp file for BZ2 decompression: /var/tmp/tmpntlqjzp9
[INFO: 2022-01-12 15:11:55 : hrit_msg] No IMPF configuration field found in prologue.
[INFO: 2022-01-12 15:11:55 : satpy.readers.utils] Using temp file for BZ2 decompression: /var/tmp/tmplypeqnc3
[WARNING: 2022-01-12 15:11:55 : satpy.readers] Don't know how to open the following files: {'/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS008___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_120___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_073___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS006___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_016___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000010___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_062___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000011___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_016___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_108___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_087___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_087___-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_097___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_073___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_134___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS008___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_097___-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_097___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS008___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000014___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_108___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_134___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_134___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_039___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_039___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_073___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS008___-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_087___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000017___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000015___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_108___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_120___-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_120___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_134___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_016___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_039___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_097___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_120___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_016___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_097___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_120___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_087___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_062___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_087___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_108___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS008___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_073___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_039___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_108___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_134___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_087___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_120___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS006___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000024___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS006___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000013___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_039___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS006___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_062___-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_062___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_134___-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_016___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS008___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS006___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_097___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000022___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_120___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000012___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_073___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_016___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000020___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_108___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_108___-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_073___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_039___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000019___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_097___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000021___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_039___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS006___-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS006___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000018___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_039___-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_073___-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000001___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_016___-000004___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_087___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS008___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_097___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_134___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_062___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_120___-000008___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_016___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS006___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_108___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_062___-000005___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000009___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-VIS008___-000002___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000016___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_087___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-HRV______-000023___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_062___-000007___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_062___-000003___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-IR_134___-000006___-202201081200-__.bz2', '/media/nas/x21308/scratch/SEVIRI/202201081200/H-000-MSG4__-MSG4________-WV_073___-000004___-202201081200-__.bz2'}
Traceback (most recent call last):
  File "/data/gholl/checkouts/protocode/mwe/seviri-bz2-failure.py", line 5, in <module>
    sc = Scene(filenames=filenames, reader=["seviri_l1b_hrit"])
  File "/data/gholl/mambaforge/envs/py39/lib/python3.9/site-packages/satpy/scene.py", line 106, in __init__
    self._readers = self._create_reader_instances(filenames=filenames,
  File "/data/gholl/mambaforge/envs/py39/lib/python3.9/site-packages/satpy/scene.py", line 127, in _create_reader_instances
    return load_readers(filenames=filenames,
  File "/data/gholl/mambaforge/envs/py39/lib/python3.9/site-packages/satpy/readers/__init__.py", line 578, in load_readers
    _check_reader_instances(reader_instances)
  File "/data/gholl/mambaforge/envs/py39/lib/python3.9/site-packages/satpy/readers/__init__.py", line 619, in _check_reader_instances
    raise ValueError("No dataset could be loaded. Either missing "
ValueError: No dataset could be loaded. Either missing requirements (such as Epilog, Prolog) or none of the provided files match the filter parameters.
```

**Environment Info:**
 - OS: openSUSE 15.3
 - Satpy Version: 0.33.2.dev32+g09e51f93 (latest git main)

**Additional context**

Indeed, https://github.com/pytroll/satpy/pull/1798 added unpacking only for prologue and epilogue, but not for the regular segments.

**Comment by pdebuyl** [comment]
Hi @gerritholl 

I started to implement this. As mentioned in #1798 the issue with image segments with respect to PRO/EPI files is the following: PRO/EPI files are read in one go whereas image files are read once upon construction and once upon loading.

I have a prototype that calls `unzip_context` twice to solve the issue, in the `__init__` and `read_band` methods of `HRITFileHandler`, at https://github.com/pdebuyl/satpy/tree/bz2_hrit_segments

I tested on an example using the "natural_colour" composite and it seems to work.

**Comment by pdebuyl** [comment]
An alternative would be to leave the temporary unzipped segment file "un-removed" on disk. This could be done by using only `unzip_file` instead of the context manager, but it would leave the files scattered around.

Another alternative would be to add an attribute pointing to the unzipped file in `HRITFileHandler` which unzips in `__init__` and removes in the finalizer `__del__`. Caveat (from <https://docs.python.org/3/reference/datamodel.html#object.__del__>)

> It is not guaranteed that [__del__()](https://docs.python.org/3/reference/datamodel.html#object.__del__) methods are called for objects that still exist when the interpreter exits.

Both alternatives create a risk of leaving large amount of data unremoved.

As a more lightweight approach that could ease a bit the `__init__` part: it is possible (in principle) to read the bzipped file without decompressing it entirely so that only reading the header part would not require to decompress the image data.

**Comment by mraspaud** [comment]
> As a more lightweight approach that could ease a bit the `__init__` part: it is possible (in principle) to read the bzipped file without decompressing it entirely so that only reading the header part would not require to decompress the image data.

Yes, it is and I like this approach. This is something I have been looking at for the Himawari HSD reader, but stumbled upon a problem: the BZ2File object can be read correctly from python function, however the `np.fromfile` and `np.memmap` functions are actually calling lower level C function that do not read the data correctly. I raised this as an issue here: https://github.com/numpy/numpy/issues/18760#
Now, a simple workaround it to use `np.frombuffer(file_object.read())` when the file is zipped, but I just wanted to make you aware of the caveat here :)

**Comment by pdebuyl** [comment]
The workaround could then be useful in the following manner:

When reading the header, use `file_object.read(size=8192)`. I checked a few files and the header is 6198 bytes long. This way, the segment does not have to be decompressed.

For the segment reading, the context manager could be used.

**Comment by pdebuyl** [comment]
PS: fixing NumPy to ingest compressed file descriptors is another way, would probably take longer :-)

**Comment by mraspaud** [comment]
> The workaround could then be useful in the following manner:
> 
> When reading the header, use `file_object.read(size=8192)`. I checked a few files and the header is 6198 bytes long. This way, the segment does not have to be decompressed.
> 
> For the segment reading, the context manager could be used.

Good idea!

**Comment by gerritholl** [comment]
Somewhat related: https://github.com/pytroll/satpy/issues/1299

**Comment by pdebuyl** [comment]
I put up a working version at the same branch:  https://github.com/pdebuyl/satpy/tree/bz2_hrit_segments

What's in?

- A new `generic_open` context manager in utils. It is like calling `with open(...) as fp` except that for filenames ending in `.bz2` it uses `bz2.open`.
- Calls to `np.fromfile` -- except for reading the bands -- are replaced by calls to `np.frombuffer(fp.read(dtype.itemsize))` (or similar). This avoid pre-loading the data as I proposed earlier as it relies on bz2 buffering mechanism.

If the branch above is considered good, I'll make a proper PR. I would be happy to have some help with writing a test as I am still not familiar with the mock library.

Bonus:
- The filename in the HRIT file handler remains the original filename, which simplifies my early work on it :-)
- As I have Himawari bz2-ed HRIT data I tested that as well and it works fine! (Data from EUMETCast).
